### ISSUE TITLE
Result compared to paper

### ISSUE THREAD

**Issue by lidaweironaldo** [donlee90/icarl#1]
Thanks for sharing the code.

What accuracy can you achieve on cifar100?

I tried to run the code, however, I cannot get anything close to the result reported in the paper.

**Comment by donlee90** [comment]
I was not able to reproduce the result either.
I haven't tested my code thoroughly, but I believe there must be some mistakes in the code.
Would love to hear about any fixes/suggestions!

**Comment by Harry-Up** [comment]
Hi, I have tried three visions of pytorch implementation for **icarl**, 
1. this version, I modified loss -> bce loss, as described in the paper
2. https://github.com/imyzx2017/icarl.pytorch, I modified parameters to the ones in the paper 
3. https://github.com/GMvandeVen/continual-learning, I changed the dataset to cifar100. In addition, this version got a good result because it takes the final task accuarcy as the incremental result, where I would modify. 

Finally, I could **not** reproduce the result in the paper. Although, the code provided by the author indeed works. I am so confused and would be happy to hear about any suggestions!

**Comment by raygoah** [comment]
Hello @Harry-Up ,
I want to know that you say the code could not reproduce the result in the paper.
It means the accuracy can't not be as high as it shows on the paper, or this code is wrong?
Is the test accuracy in the training too low or the testing accuracy is too low after completing training?

Thank you very much.

**Comment by arthurdouillard** [comment]
Hello @Harry-Up:

The second pytorch implementation that you've tried is an anonymous fork of my code actually (not very cool of them to do it this way).
You can find a more updated version here: https://github.com/arthurdouillard/incremental_learning.pytorch

What parameters were different according to you from the paper ? I've tried to respect them.

Btw in my implem, I'm plotting in the graph, the "average incremental accuracy" instead of the "task accuracy" and get the *exact* same curve as in the paper. But I do not understand what is the 64.1% results from the tables. Do you know something about it?

**Comment by Harry-Up** [comment]
Hello, @raygoah :

Not exactly. The result could be reproduced by the code from the author. However, it is based on tensorflow / TheanoLasagne, and I am looking for a pytorch implementation.

**Comment by Harry-Up** [comment]
Hello, @arthurdouillard :

Thank you for your elegant implementation. 

I ran this code and an error is raised. It is because the size of the feature map before the avgpool layer is 2*2, while the kernel size of avgpool layer is set as 4*4. So, I modified the kernel size as 2*2. It seems well. 

However, the result is not as high as plotted. I got [0.846， 0.72，0.626，0.552，0.486，0.463，0.43，0.399，0.384，0.366] for 10 stage tasks with resnet18. I have run the resnet34 with the same setting, the result also went down to 0.36. 

My system is ubuntu 18.04, python 3.7 and pytorch 1.0.0. Look forward to hearing any suggestion.

**Comment by arthurdouillard** [comment]
@Harry-Up:
 
I'm not sure why you got this error, all of my code is committed :/ I'll look into it.

Indeed the result averaged over 40 runs are: [0.83, 0.73, 0.63, 0.56, 0.5, 0.46, 0.43, 0.41, 0.38, 0.36].
However if I plot the the average incremental accuracy instead of the accuracy, my results are [0.83, 0.78, 0.73, 0.69, 0.65, 0.62, 0.59, 0.57, 0.55, 0.53]. Which are the same results that are displayed on the figure 2. of iCaRL.

Am I making a mistake to use the average incremental accuracy? It's quite amazing that my curve can be the same as theirs doing so. But I'm cannot explain then the 64.1% displayed in the table 1.

If you know anything how to improve my implementation, please tell me or make a MR. I'd be glad to have your input.

**Comment by Harry-Up** [comment]
Hello, @arthurdouillard :

I am sorry to make a misunderstanding about _average incremental accuracy_, especially it is a recognized indicator. I thought before that it would be the mean of all task accuracies at the current task, and it should be the mean of all task accuracies until now.

![image](https://user-images.githubusercontent.com/26539945/57290001-1d257700-70ef-11e9-9a7e-f54d60f83d78.png)

In addition, for the avgpool problem, there are 4 times dimension halving, including maxpool, layer2, layer3 and layer4. So the image size will be from $32 \times 32$ to $2 \times 2$, and it will not be suitable for AvgPool2d(4).

As for 64.1%, I have no idea too. But, it seems to be the mean of average incremental accuracy again. Is it?

**Comment by arthurdouillard** [comment]
@Harry-Up 
Yes, about the average incremental accuracy: you see that my first results ([0.83, 0.73, 0.63, 0.56, 0.5, 0.46, 0.43, 0.41, 0.38, 0.36]), when each value is averaged with its previous values I get the second results ([0.83, 0.78, 0.73, 0.69, 0.65, 0.62, 0.59, 0.57, 0.55, 0.53]).  But this is far from 64.1%.

I'll return to you when the author answer me about what results is using average incremental accuracy. 

You're right about the avgpool, it's weird that it didn't break on my side. I've fixed it thanks.

**Comment by Harry-Up** [comment]
@arthurdouillard :

I mean (0.83+0.78+0.73+0.69+0.65+0.62+0.59+0.57+0.55+0.53)/10 = 65.4%, it is similar to 64.1%

**Comment by arthurdouillard** [comment]
Oh you're right. So an average of an average incremental accuracy.

The results are indeed close. I'm waiting for the author to answer me about it.

Have you understood from author code in theano which metric was used?

EDIT: Looking at the figure 2. results & making an average of it lead also to 64.1%. That must be it.

**Comment by raygoah** [comment]
Hello @arthurdouillard :

First, thanks for your awesome work.
But, I have some questions.
Why we need to average the value now and the previous ones to get the average accuracy?
Shouldn't the value in your first result each means the accuracy of model doing classify to the current classes?
Then, I wonder what is the meaning of your second result (average incremental accuracy) ?

Thank you very much.

**Comment by arthurdouillard** [comment]
@raygoah: From what I understand with a conversation with the author a week ago:
Simply using the accuracy is not enough in some cases. Example:

Two models: M1, and M2.
M1 has accuracy [0.5, 0.4, 0.3] & average incremental accuracy [0.5, 0.45, 0.4]
M2 has accuracy [0.8, 0.6, 0.3] & average incremental accuracy [0.8, 0.7, 0.56]

While last accuracy of M1 & M1 are equal, the two models are clearly not equal, M2 is better as the average incremental accuracy show it!

Hope that helps.

**Comment by raygoah** [comment]
@arthurdouillard : 

Your explanation really makes sense. 
I agree the average incremental accuracy is a great indicator.
I think maybe I misunderstood the meaning of Figure 2 in the paper before.
Are the images in the Figure 2 in the icarl paper also the average accuracy, right?

Thank you very much.

**Comment by arthurdouillard** [comment]
About this, I'm not sure at 100%. My results would say yes, but I've asked the author to confirm just to be sure.

**Comment by arthurdouillard** [comment]
The author hasn't respond to me, but I'm pretty sure I was wrong. The plot results are the accuracy was the table results are the averages.

**Comment by Harry-Up** [comment]
Yeah, I would like to report the final incremental learning result as the
average of "average incremental accuracy", is it?

Meanwhile, I plan to fix some problems of the LwF code to make it really
works, referring to  https://github.com/GMvandeVen/continual-learning.

Good luck

Arthur Douillard <notifications@github.com> 于2019年5月24日周五 下午11:52写道：

**Comment by arthurdouillard** [comment]
I was wrong: each curve's dot is the accuracy from 0 to current class. The final number is the average of these.
You can find in my repo (https://github.com/arthurdouillard/incremental_learning.pytorch) a working implementation of iCaRL where with the right parameters I reach an avg of 64.34.

**Comment by suanrong** [comment]
> I was wrong: each curve's dot is the accuracy from 0 to current class. The final number is the average of these.
> You can find in my repo (https://github.com/arthurdouillard/incremental_learning.pytorch) a working implementation of iCaRL where with the right parameters I reach an avg of 64.34.

Hi, I run your code and got an avg of 60.68. I tried different parameters but I couldn't reach 64.34.
Would you mind sharing you parameters of iCaRL? Thanks a lot ~

**Comment by arthurdouillard** [comment]
My code has changed a lot (both the public and my private) version. But try those parameters:
```
lr: 2.
weight_decay: 0.00005
scheduling:
  - 49
  - 63
lr_decay: 0.2
optimizer: sgd
epochs: 70
```

**Comment by suanrong** [comment]
> My code has changed a lot (both the public and my private) version. But try those parameters:
> 
> ```
> lr: 2.
> weight_decay: 0.00005
> scheduling:
>   - 49
>   - 63
> lr_decay: 0.2
> optimizer: sgd
> epochs: 70
> ```

Thanks you!
I have got the same result reported in paper for 5,10,20 and 50 classes per batch. However, when I try 2 classes per batch, the result is extremely bad.  The accuracy finally reach at about 20% while it is about 45% reported in paper.

Do you have any idea about that?

**Comment by arthurdouillard** [comment]
Not sure it's been a while.

If the results are that bad, it may be a gradient explosion? Try to clip the gradients value between values like (-5, +5) (or less).

**Comment by suanrong** [comment]
> clip the gradients value between values like (-5, +5) (or less)

Thanks，I will try.

**Comment by KOuldamer** [comment]
Hello ,

Thanks for sharing the code.
total_classes = 10
num_classes = 10
for s in range(0, total_classes, num_classes): 
          S=0 Why you make this loop?
